File: processVolumeData.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <fstream>
#include <teem/nrrd.h>

struct Grid {
    std::vector<float> data;
    int nx, ny, nz;
};

Grid load_nrrd_data(const std::string& file_path) {
    Nrrd* nrrd = nrrdNew();
    if (nrrdLoad(nrrd, file_path.c_str(), NULL)) {
        char *err = biffGetDone(NRRD);
        std::cerr << "Error reading NRRD file: " << err << std::endl;
        free(err);
        nrrdNuke(nrrd);
        exit(1);
    }

    size_t total_size = nrrdElementNumber(nrrd);

    std::vector<float> data(total_size);

    float* data_ptr = static_cast<float*>(nrrd->data);

    std::copy(data_ptr, data_ptr + total_size, data.begin());

    int nx = nrrd->axis[0].size;
    int ny = nrrd->axis[1].size;
    int nz = nrrd->axis[2].size;

    nrrdNuke(nrrd); // Properly dispose of the Nrrd structure

    std::ofstream out_file("output_data.txt");
    for (const auto& value : data) {
        out_file << value << std::endl;
    }
    out_file.close();


    return {data, nx, ny, nz};
}

bool is_cube_active(const Grid& grid, int x, int y, int z, float isovalue) {
    // Define edges and check for bipolar characteristics
    std::vector<std::pair<int, int>> edges = {
        {0, 1}, {1, 2}, {2, 3}, {3, 0}, // Bottom face edges
        {4, 5}, {5, 6}, {6, 7}, {7, 4}, // Top face edges
        {0, 4}, {1, 5}, {2, 6}, {3, 7}  // Vertical edges
    };

    // Edge-to-vertex mapping for cube
    std::vector<std::tuple<int, int, int>> vertex_offsets = {
        {0, 0, 0}, {1, 0, 0}, {1, 1, 0}, {0, 1, 0}, // Bottom face
        {0, 0, 1}, {1, 0, 1}, {1, 1, 1}, {0, 1, 1}  // Top face
    };

    //std::cout << "Checking cube at (" << x << ", " << y << ", " << z << "):\n"; // Debugging output
    for (const auto& edge : edges) {
        auto [v1x, v1y, v1z] = vertex_offsets[edge.first];
        auto [v2x, v2y, v2z] = vertex_offsets[edge.second];

        int idx1 = (x + v1x) + (y + v1y) * grid.nx + (z + v1z) * grid.nx * grid.ny;
        int idx2 = (x + v2x) + (y + v2y) * grid.nx + (z + v2z) * grid.nx * grid.ny;

        if (idx1 < grid.data.size() && idx2 < grid.data.size()) { // Check to ensure indices are within bounds
            float val1 = grid.data[idx1];
            float val2 = grid.data[idx2];
            //std::cout << "Edge (" << edge.first << ", " << edge.second << ") values: " << val1 << ", " << val2 << "\n"; // Debugging output


            if ((val1 < isovalue && val2 >isovalue) || (val1 > isovalue && val2 < isovalue)) {
                //std::cout << "Active edge detected, cube is active.\n"; // Debugging output
                return true;
            }
        }

    }
    //std::cout << "No active edges, cube is inactive.\n"; // Debugging output
    return false;
}

std::vector<std::tuple<float, float, float>> find_active_cubes(const Grid& grid, float isovalue) {
    std::vector<std::tuple<float, float, float>> centers;
    for (int i = 0; i < grid.nx - 1; ++i) {
        for (int j = 0; j < grid.ny - 1; ++j) {
            for (int k = 0; k < grid.nz - 1; ++k) {
                if (is_cube_active(grid, i, j, k, isovalue)) {
                    centers.emplace_back(i + 0.5, j + 0.5, k + 0.5);
                    //std::cout << "Center added: (" << i + 0.5 << ", " << j + 0.5 << ", " << k + 0.5 << ")\n\n"; // Debugging output
                }
            }
        }
    }
    return centers;
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <file_path> <isovalue>\n";
        return 1;
    }

    int count = 0;

    std::string file_path = argv[1];
    float isovalue = std::stof(argv[2]);
    
    Grid grid = load_nrrd_data(file_path);

    std::cout << "Data size: " << grid.nx << ", " << grid.ny << ", " << grid.nz << std::endl;

    std::cout << "Raw data loaded:\n";
    for (int k = 0; k < grid.nz; ++k) {
        for (int j = 0; j < grid.ny; ++j) {
            for (int i = 0; i < grid.nx; ++i) {
                //std::cout << grid.data[i + j * grid.nx + k * grid.nx * grid.ny] << " ";
            }
            //std::cout << std::endl;
        }
        //std::cout << std::endl;
    }

    auto centers = find_active_cubes(grid, isovalue);

    std::cout << "Centers of active cubes:\n";
    if (centers.empty()) {
        std::cout << "No active cubes found.\n";
    } else {
        for (auto& center : centers) {
            ++count;
            //std::cout << "(" << std::get<0>(center) << " " << std::get<1>(center) << " " << std::get<2>(center) << ")" << std::endl;
        }
        std::cout << "Total number of centers of active cubes found: " << count << std::endl;
    }

    return 0;
}

File: processVolumeData.py
import numpy as np
import nibabel as nib
import os
import sys

def parse_nhdr_header(filename):
    """
    Parse the NHDR header file to extract metadata about the scalar grid.

    Parameters:
        filename (str): The path to the NHDR file.

    Returns:
        dict: A dictionary containing the metadata from the NHDR file.
    """
    header = {}
    with open(filename, 'r') as file:
        for line in file:
            if ':' in line:
                key, value = line.strip().split(':')
                header[key.strip()] = value.strip()
    return header

def load_raw_data(header, raw_file_path):
    """
    Load the raw binary data from a file based on the NHDR metadata.

    Parameters:
        header (dict): Metadata containing data type and dimensions.
        raw_file_path (str): Path to the raw data file.

    Returns:
        np.array: A NumPy array containing the structured raw data.
    """
    dtype_map = {'unsigned char': np.uint8, 'short': np.int16, 'float': np.float32}
    data_type = dtype_map[header['type']]
    dimensions = tuple(map(int, header['sizes'].split()))
    with open(raw_file_path, 'rb') as file:
        data = np.fromfile(file, dtype=data_type).reshape(dimensions)
    return data

def load_nhdr_data(nhdr_path):
    """
    Load the raw binary data for NHDR files.

    Parameters:
        nhdr_path (str): Path to the NHDR file.

    Returns:
        np.array: A NumPy array containing the structured raw data.
    """
    header = parse_nhdr_header(nhdr_path)
    raw_file_path = os.path.join(os.path.dirname(nhdr_path), header['data file'])
    return load_raw_data(header, raw_file_path)


def read_nrrd_header(file_path):
    """Reads the header of an NRRD file to determine data dimensions and type."""
    header = {}
    with open(file_path, 'r') as file:
        for line in file:
            if line.strip() == '':
                break
            if ':' in line:
                key, value = line.strip().split(':', 1)
                header[key.strip()] = value.strip()
    return header

def read_nrrd_data(file_path):
    """Reads ASCII encoded NRRD data from a given file."""
    header = read_nrrd_header(file_path)
    dimensions = tuple(map(int, header['sizes'].split()))
    dtype_map = {
        'unsigned char': np.uint8,
        'short': np.int16,
        'float': np.float32
    }
    data_type = dtype_map[header['type']]
    
    data = []
    with open(file_path, 'r') as file:
        # Skip the header lines
        while True:
            line = file.readline()
            if line.strip() == '':  # Empty line indicates end of header
                break
        # Read the data
        for line in file:
            data.extend(map(float, line.split()))

    print(np.array(data).reshape(dimensions).astype(data_type))
    
    return np.array(data).reshape(dimensions).astype(data_type)
    
def load_data(file_path):
    if file_path.endswith('.nhdr'):
        return load_nhdr_data(file_path)
    elif file_path.endswith('.nrrd'):
        return read_nrrd_data(file_path)
    else:
        raise ValueError("Unsupported File Format")
    
def find_active_cubes(grid, isovalue):
    """
    Identify active cubes within the scalar grid that contain at least one bipolar edge.

    Parameters:
        grid (np.array): The 3D array of scalar values.
        isovalue (float): The threshold value to determine bipolar edges.

    Returns:
        list: A list of tuples representing the indices of active cubes.
    """
    nx, ny, nz = grid.shape
    active_cubes = []
    for i in range(nx - 1):
        for j in range(ny - 1):
            for k in range(nz - 1):
                if is_cube_active(grid, i, j, k, isovalue):
                    active_cubes.append((i, j, k))
    return active_cubes

def is_cube_active(grid, x, y, z, isovalue):
    """
    Determines if a cube has at least one bipolar edge.
    A bipolar edge is one where the scalar values at its endpoints straddle the isovalue.

    Parameters:
        grid (np.array): The 3D array of scalar values.
        x, y, z (int): Indices of the lower corner of the cube in the grid.
        isovalue (float): The threshold value for determining bipolar edges.

    Returns:
        bool: True if at least one edge is bipolar, False otherwise.
    """
    # Edges are defined by pairs of indices into the cube's corners
    # There are 12 edges in a cube
    edges = [
        ((x, y, z), (x + 1, y, z)),
        ((x, y, z), (x, y + 1, z)),
        ((x, y, z), (x, y, z + 1)),
        ((x + 1, y + 1, z), (x, y + 1, z)),
        ((x + 1, y + 1, z), (x + 1, y, z)),
        ((x + 1, y + 1, z), (x + 1, y + 1, z + 1)),
        ((x, y + 1, z + 1), (x, y, z + 1)),
        ((x, y + 1, z + 1), (x, y + 1, z)),
        ((x, y + 1, z + 1), (x + 1, y + 1, z + 1)),
        ((x + 1, y, z + 1), (x + 1, y, z)),
        ((x + 1, y, z + 1), (x, y, z + 1)),
        ((x + 1, y, z + 1), (x + 1, y + 1, z + 1))
    ]

    # Check each edge to determine if it is bipolar
    for (x0, y0, z0), (x1, y1, z1) in edges:
        val0 = grid[x0, y0, z0]
        val1 = grid[x1, y1, z1]
        if (val0 < isovalue and val1 > isovalue) or (val0 > isovalue and val1 < isovalue):
            return True  # This edge is bipolar

    return False  # No bipolar edges found


def process_volume_data(file_path, isovalue, output_file_path):
    """
    Process the volume data from NHDR or NRRD file, identify active cubes, compute centers, and write them to a file.

    Parameters:
        file_path (str): Path to the NHDR/NRRD file.
        isovalue (float): The isovalue to identify active cubes.
        output_file_path (str): Path to the output file where centers will be written.
    """
    grid = load_data(file_path)
    print("Raw data size:")
    print(grid.shape)
    
    active_cubes = find_active_cubes(grid, isovalue)
    centers = [(x + 0.5, y + 0.5, z + 0.5) for x, y, z in active_cubes]
    
    # Writing centers to file
    with open(output_file_path, "w") as file:
        for center in centers:
            file.write(f"{center[0]} {center[1]} {center[2]}\n")
    
    print("Number of active cubes:")
    print(len(centers))

def main():
    if len(sys.argv) < 3:
        print("Usage: python processVolumeData.py <file_path> <isovalue> <output_file_path>\n")
        sys.exit(1)
    
    file_path = sys.argv[1]
    isovalue = float(sys.argv[2])
    output_file_path = sys.argv[3]
    process_volume_data(file_path, isovalue, output_file_path)

if __name__ == "__main__":
    main()
